src/Evolution/Systems/GeneralizedHarmonic/CharFields.cpp

```diff
@@ -0,0 +1,132 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+// IWYU pragma: no_include <array>
+
+#include "CharFields.hpp"
+#include "DataStructures/DataBox/Prefixes.hpp" // IWYU pragma: keep
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp" // IWYU pragma: keep
+#include "Evolution/Systems/GeneralizedHarmonic/Tags.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/MakeWithValue.hpp"
+
+// IWYU pragma: no_forward_declare Tensor
+// IWYU pragma: no_forward_declare Tags::CharSpeed
+
+namespace GeneralizedHarmonic {
+
+template <size_t Dim, typename Frame>
+typename Tags::EvolvedFieldsFromCharacteristic<Dim, Frame>::type
+EvolvedFieldsFromCharacteristicCompute<Dim, Frame>::function(
+    const Scalar<DataVector>& gamma_2,
+    const tnsr::aa<DataVector, Dim, Frame>& u_psi,
+    const tnsr::iaa<DataVector, Dim, Frame>& u_zero,
+    const tnsr::aa<DataVector, Dim, Frame>& u_plus,
+    const tnsr::aa<DataVector, Dim, Frame>& u_minus,
+    const tnsr::i<DataVector, Dim, Frame>& normal_one_form) noexcept {
+  auto evolved_fields =
+      make_with_value<typename Tags::EvolvedFieldsFromCharacteristic<Dim,
+          Frame>::type>(get(gamma_2), 0.);
+
+  for (size_t mu = 0; mu < Dim + 1; ++mu) {
+    for (size_t nu = 0; nu < mu + 1; ++nu) {
+      get<::Tags::Field<Tags::Psi<Dim, Frame>>>(evolved_fields).get(mu, nu) =
+                                    u_psi.get(mu, nu);
```

Comment:
Is the loop necessary? Can't you initialize the evolved field tensors directly?

Reply:
done